CSS Custom Highlights optimization add flag for when live ranges are changed

jesxilin opened the pull request:
#### 21b682820f4a338153863fdb544d2fc12a1daf14
<pre>
CSS Custom Highlights optimization add flag for when live ranges are changed
<a href="https://bugs.webkit.org/show_bug.cgi?id=259804">https://bugs.webkit.org/show_bug.cgi?id=259804</a>
rdar://113361192

Reviewed by Ryosuke Niwa.

Does not collect range data for live ranges that were not changed/updated.
Before, would collect range data for all live ranges all the time, and repainting them.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::collectRangeDataFromRegister):
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::setStart):
(WebCore::Range::setEnd):
(WebCore::Range::nodeChildrenChanged):
(WebCore::Range::nodeChildrenWillBeRemoved):
(WebCore::Range::nodeWillBeRemoved):
(WebCore::Range::textInserted):
(WebCore::Range::textRemoved):
(WebCore::Range::textNodesMerged):
(WebCore::Range::textNodeSplit):
* Source/WebCore/dom/Range.h:

Canonical link: <a href="https://commits.webkit.org/266614@main">https://commits.webkit.org/266614@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1731559070fc6f1af45077d377575caf361cd9e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14255 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14568 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14908 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15990 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13499 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14378 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17081 "Failed to compile WebKit") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14646 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16171 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14432 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/23/builds/17081 "Failed to compile WebKit") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12098 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16711 "Built successfully") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/23/builds/17081 "Failed to compile WebKit") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12859 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19866 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/23/builds/17081 "Failed to compile WebKit") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13023 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16221 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13564 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11418 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12847 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17184 "Built successfully") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1699 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13409 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2989,15 +2989,28 @@ void Document::collectRangeDataFromRegister(Vector<WeakPtr<HighlightRangeData>>&
     for (auto& highlight : highlightRegister.map()) {
         for (auto& rangeData : highlight.value->rangesData()) {
-            // FIXME: For live ranges, we can optimize by only performing this when the range changed.
+            // Skip if positions are not null for static range.
```
This comment just repeats what the code already conveys. Pease remove.
Alternatively, add "why" comment instead of "what" comment.

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2989,15 +2989,28 @@ void Document::collectRangeDataFromRegister(Vector<WeakPtr<HighlightRangeData>>&
             if (rangeData->startPosition().isNotNull() && rangeData->endPosition().isNotNull() && !rangeData->range().isLiveRange())
                 continue;
+
+            // Skip if a live range did not change start or end.
```
Ditto. This comment doesn't add any information beyond what the code already tells us.

rniwa commented on Source/WebCore/dom/Range.h:
```diff
@@ -52,6 +52,9 @@ class Range final : public AbstractRange, public CanMakeWeakPtr<Range> {
     WEBCORE_EXPORT Node* commonAncestorContainer() const;
 
+    void resetDidChange() { m_didChange = false; }
+    bool didChange() const { return m_didChange; }
```
These functions are specific for highlights so perhaps we should communicate that.
e.g. resetDidChangeHighlight & didChangeHighlight.

jesxilin replied:
Got it, will change!

jesxilin committed:
CSS Custom Highlights optimization add flag for when live ranges are changed
https://bugs.webkit.org/show_bug.cgi?id=259804
rdar://113361192

Reviewed by Ryosuke Niwa.

Does not collect range data for live ranges that were not changed/updated.
Before, would collect range data for all live ranges all the time, and repainting them.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::collectRangeDataFromRegister):
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::setStart):
(WebCore::Range::setEnd):
(WebCore::Range::nodeChildrenChanged):
(WebCore::Range::nodeChildrenWillBeRemoved):
(WebCore::Range::nodeWillBeRemoved):
(WebCore::Range::textInserted):
(WebCore::Range::textRemoved):
(WebCore::Range::textNodesMerged):
(WebCore::Range::textNodeSplit):
* Source/WebCore/dom/Range.h:

Canonical link: https://commits.webkit.org/266614@main
